mflowgen/common/scripts/fix-shorts.tcl

```diff
@@ -0,0 +1,132 @@
+verify_drc -layer_range { M4 M9 }
+
+
+
+########################################################################
+# Fix shorts 2: Identify the shorts
+set shorts [dbGet top.markers { .subType eq "Metal Short" }]
+if { $shorts == "0x0" } {
+    echo "@file_info: No shorts found"
+} {
+   # Fix shorts 3: Show how many shorts were found
+    set nshorts [llength $shorts]
+    echo "@file_info: Found $nshorts short circuit(s):"
+    echo [ dbGet [dbGet top.markers { .subType eq "Metal Short" }].message ]
+
+    # Fix shorts 4: See if globalDetailRoute can fix the shorts with eco
+    echo "@file_info: Fixing short circuits"
+    setNanoRouteMode -routeWithEco true
+    setNanoRouteMode -drouteEndIteration 2
+    globalDetailRoute
+
+    # Fix shorts 5: Check your work
+    clearDrc
+    verify_drc -layer_range { M4 M9 }
+
+    set shorts [dbGet top.markers { .subType eq "Metal Short" }]
+    if { $shorts == "0x0" } {
+        echo "@file_info: All shorts fixed"
+    } {
+
+        # PE requires two tries maybe
+        echo "@file_info: Oops looks like I failed; lemme try again"
+
+        # Fix shorts 3a: Show how many shorts were found
+        set nshorts [llength $shorts]
+        echo "@file_info: (2)Found $nshorts short circuit(s):"
+        echo [ dbGet [dbGet top.markers { .subType eq "Metal Short" }].message ]
+
+        # Fix shorts 4a: See if globalDetailRoute can fix the shorts with eco
+        echo "@file_info: (2)Fixing short circuits"
+        setNanoRouteMode -routeWithEco true
+        setNanoRouteMode -drouteEndIteration 2
+        globalDetailRoute
+
+        # Fix shorts 5a: Check your work
+        clearDrc
+        verify_drc -layer_range { M4 M9 }
+
+        set shorts [dbGet top.markers { .subType eq "Metal Short" }]
+        if { $shorts == "0x0" } {
+            echo "@file_info: All shorts fixed"
+
+            # FIXME doesn't the wrapper script save the design?
+            # FIXME in which case we don't need this line here?
+            # Save changes *only* if shorts fixed
+            saveDesign checkpoints/design.checkpoint/save.enc -user_path
+
+        } {
+            echo "@file_info: Oops looks like I failed again oh no"
+            echo ""
+            echo "**ERROR: Metal shorts exist, see log for details"
+            echo "@file_info: Found $nshorts short circuit(s):"
+            echo "@file_info: Giving gup now"
+        }
+    }
+}
+
+########################################################################
+# Restore route blockages
+redraw; sleep 1
+echo "@file_info: Restoring $nblockages route blockages"
+for { set i 0 } { $i < $nblockages } { incr i } {
+    printf "%2d " $i
+    echo createRouteBlk \
+        -layer $blockage_layers($i) \
+        -box $blockage_boxes($i)    \
+        -name $blockage_names($i)   \
+        ;
+    createRouteBlk \
+        -layer $blockage_layers($i) \
+        -box $blockage_boxes($i)    \
+        -name $blockage_names($i)   \
+        ;
+}
+redraw; sleep 1
```

Comment:
Just a general comment on this script... I really like the idea of having an automated way of fixing shorts in innovus! It seems like this could be fairly easily extended to fix all metal geometry DRCs, not just shorts. That can be left for a future PR though.

Reply:
Agreed. But I think it's crazy that we have to do this. Innovus should be smarter. And it definitely shouldn't leave us with shorted wires without throwing some kind of error or at least a warning!!